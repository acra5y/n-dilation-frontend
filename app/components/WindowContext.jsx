import React, { useContext } from "react";

const WindowContext = React.createContext();

export const useWindowContext = () => useContext(WindowContext);

export default WindowContext;
