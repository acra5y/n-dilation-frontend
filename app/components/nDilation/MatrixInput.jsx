import React, { useState } from "react";
import styled from "styled-components";

const darkGreen = "#4c8f11";

const StyledForm = styled.form`
    display: flex;
    flex-direction: column;
    align-items: center;
`;
StyledForm.displayName = "StyledForm";

const TextareaWrapper = styled.div`
    background-color: ${darkGreen};
    padding: 2px;
    border-radius: 3px;
    box-shadow: 0 0 2px 2px ${darkGreen};
`;
TextareaWrapper.displayName = "TextareaWrapper";

const StyledTextarea = styled.textarea`
    border: 2px solid #cde843;
    border-radius: 3px;

    &:hover,
    &:focus {
        border-color: #fae56b;
    }
`;
StyledTextarea.displayName = "StyledTextarea";

const StyledInput = styled.input`
    border-radius 18px;
    border: 2px solid ${darkGreen};
    box-shadow: 0 0 1px 1px ${darkGreen};
    padding 2px 7px;
`;
StyledInput.displayName = "StyledInput";

const matrixRegex = /^(([+-]?\d+(\.\d+)?),)*([+-]?\d+(\.\d+)?)$/;
const isMatrix = matrixRegex.test.bind(matrixRegex);
const isSquare = number => number > 0 && Math.sqrt(number) % 1 === 0;

const MatrixInput = ({ onSubmit }) => {
    const [input, setInput] = useState("");

    return (
        <StyledForm
            onSubmit={ev => {
                ev.preventDefault();

                if (isMatrix(input.replace(/\s/g, ""))) {
                    const parsed = JSON.parse(`[${input}]`);

                    if (isSquare(parsed.length)) {
                        onSubmit(parsed);
                    }
                }
            }}
        >
            <label htmlFor="matrix-input">Enter Matrix</label>
            <TextareaWrapper>
                <StyledTextarea
                    name="matrix-input"
                    type="text"
                    onChange={ev => setInput(ev.target.value)}
                    autoFocus
                    rows={2}
                    placeholder="0, 0.5, 0, 0"
                />
            </TextareaWrapper>
            <StyledInput type="submit" value="Compute" />
        </StyledForm>
    );
};

export default MatrixInput;
