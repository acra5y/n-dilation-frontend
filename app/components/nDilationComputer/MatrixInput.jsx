import React, { useState } from "react";

const MatrixInput = ({ onSubmit }) => {
    const [input, setInput] = useState("");

    return (
        <form
            onSubmit={ev => {
                ev.preventDefault();
                onSubmit(input);
            }}
        >
            <label htmlFor="matrix-input">Enter Matrix</label>
            <textarea
                name="matrix-input"
                type="text"
                onChange={ev => setInput(ev.target.value)}
            />
            <input type="submit" value="Compute" />
        </form>
    );
};

export default MatrixInput;
