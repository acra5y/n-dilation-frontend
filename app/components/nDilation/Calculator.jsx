import React, { useState } from "react";

import fetchNDilation from "../../lib/fetchNDilation";
import { useWindowContext } from "../WindowContext";
import MatrixInput from "./MatrixInput";

const createOnSubmitHandler = (
    fetch,
    setDilation,
    setError
) => async matrix => {
    try {
        const response = await fetchNDilation(fetch, matrix);

        if (response.ok) {
            const body = await response.json();
            setDilation(body.value);
        }
    } catch (e) {
        setDilation(null);
        setError(e);
    }
};

const Calculator = () => {
    const [dilation, setDilation] = useState(null);
    const [error, setError] = useState(null);
    const window = useWindowContext();

    return (
        <div>
            <MatrixInput
                onSubmit={
                    window &&
                    createOnSubmitHandler(window.fetch, setDilation, setError)
                }
            />
            {dilation && <div>{JSON.stringify(dilation)}</div>}
            {error && <div>Ooops, something went terribly wrong ðŸ¤¯</div>}
        </div>
    );
};

export default Calculator;
