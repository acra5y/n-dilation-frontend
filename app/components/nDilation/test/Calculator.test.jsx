import React from "react";
import { shallow } from "enzyme";

import MatrixInput from "../MatrixInput";
import * as WindowContext from "../../WindowContext";
import Calculator from "../Calculator";

const createWindow = (
    response = {
        ok: true,
        json: () => Promise.resolve({}),
    }
) => ({
    fetch: jest.fn(() => Promise.resolve(response)),
});
const render = (window = {}) => {
    jest.spyOn(WindowContext, "useWindowContext").mockImplementation(
        () => window
    );
    return shallow(<Calculator />);
};

describe("Calculator", () => {
    it("should render", () => {
        jest.spyOn(WindowContext, "useWindowContext").mockImplementation(
            () => ({})
        );
        const component = render();

        expect(component).toMatchSnapshot();
    });

    describe("on client", () => {
        it("call fetch when onSubmit on MatrixInput is called", async () => {
            const window = createWindow();
            const component = render(window);

            await component.find(MatrixInput).prop("onSubmit")();

            expect(window.fetch.mock.calls.length).toEqual(1);
        });

        it("should render the dilation after is has been fetched", async () => {
            const response = {
                ok: true,
                json: () => Promise.resolve({ value: [1, 2, 3, 4] }),
            };
            const component = render(createWindow(response));
            await component.find(MatrixInput).prop("onSubmit")();
            expect(component).toMatchSnapshot();
        });

        it("should not throw an exception if the fetch is not ok", async () => {
            const response = {
                ok: false,
            };
            const component = render(createWindow(response));
            await component.find(MatrixInput).prop("onSubmit")();
            expect(component).toMatchSnapshot();
        });
    });
});
