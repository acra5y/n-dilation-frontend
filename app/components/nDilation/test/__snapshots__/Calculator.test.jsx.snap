// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator on client should catch an exception thrown by fetch 1`] = `
<div>
  <MatrixInput
    onSubmit={[Function]}
  />
  <div>
    Ooops, something went terribly wrong ðŸ¤¯
  </div>
</div>
`;

exports[`Calculator on client should not throw an exception if the fetch is not ok 1`] = `
<div>
  <MatrixInput
    onSubmit={[Function]}
  />
</div>
`;

exports[`Calculator on client should render the dilation after is has been fetched 1`] = `
<div>
  <MatrixInput
    onSubmit={[Function]}
  />
  <div>
    [1,2,3,4]
  </div>
</div>
`;

exports[`Calculator should render 1`] = `
<div>
  <MatrixInput
    onSubmit={[Function]}
  />
</div>
`;
